File: frontend/src/Components/Forms/EditPictureForm.js
import React, { useEffect, useState } from "react";
import server from "../Static/Constants";
import { useLocation } from "react-router-dom";

const EditPictureForm = () => {
    const [picture, setPicture] = useState(null);
    const [error, setError] = useState(null);
    const [formData, setFormData] = useState({});
    const [successMessage, setSuccessMessage] = useState("");
    const location = useLocation();
    const search = location.search;

    useEffect(() => {
        const fetchPicture = async () => {
            try {
                const token = localStorage.getItem("token");
                const params = new URLSearchParams(search);
                const serial = params.get("serial");

                if (!serial) {
                    setError("Picture parameter is missing");
                    return;
                }

                const response = await fetch(`${server}/get/picture_record/${serial}`, {
                    method: "GET",
                    headers: {
                        Authorization: token,
                        "Content-Type": "application/json",
                    },
                });

                if (!response.ok) {
                    throw new Error("Failed to fetch picture data.");
                }

                const responseData = await response.json();

                if (responseData.data) {
                    setPicture(responseData.data);
                    setFormData({
                        title: responseData.data.picture_title || "",
                        description: responseData.data.picture_description || "",
                        tags: responseData.data.picture_tags ? responseData.data.picture_tags.split(",") : [],
                        people: responseData.data.picture_people ? responseData.data.picture_people.split(",") : [],
                    });
                } else {
                    setError("No picture found.");
                }
            } catch (error) {
                setError(error.message);
            }
        };

        fetchPicture();
    }, [search]);

    const handleSubmit = async (e) => {
        e.preventDefault();
        setSuccessMessage("");
        setError("");

        try {
            const token = localStorage.getItem("token");
            const serial = new URLSearchParams(search).get("serial");

            if (!serial) {
                setError("Invalid request. No picture serial found.");
                return;
            }

            const response = await fetch(`${server}/update/picture_record/${serial}/`, {
                method: "PATCH",
                headers: {
                    Authorization: token,
                    "Content-Type": "application/json",
                },
                body: JSON.stringify({
                    picture_title: formData.title,
                    picture_description: formData.description,
                    tags: formData.tags.join(","),
                    people: formData.people.join(","),
                }),
            });

            if (!response.ok) {
                throw new Error("Failed to update picture details.");
            }

            setSuccessMessage("Picture details updated successfully!");
        } catch (error) {
            setError(error.message);
        }
    };

    const handleChange = (e, field, index) => {
        const updatedData = { ...formData };
        updatedData[field][index] = e.target.value;
        setFormData(updatedData);
    };

    const handleRemoveField = (field, index) => {
        const updatedData = { ...formData };
        updatedData[field].splice(index, 1);
        setFormData(updatedData);
    };

    const handleAddField = (field) => {
        const updatedData = { ...formData };
        updatedData[field].push("");
        setFormData(updatedData);
    };

    return (
        <div>
            {error && <p style={{ color: "red" }}>{error}</p>}
            {successMessage && <p style={{ color: "green" }}>{successMessage}</p>}
            {picture && (
                <>
                    <form onSubmit={handleSubmit}>
                        <label>
                            Title:
                            <input
                                type="text"
                                name="title"
                                value={formData.title}
                                onChange={(e) => setFormData({ ...formData, title: e.target.value })}
                            />
                        </label>
                        <label>
                            Description:
                            <input
                                type="text"
                                name="description"
                                value={formData.description}
                                onChange={(e) => setFormData({ ...formData, description: e.target.value })}
                            />
                        </label>

                        {/* Tags Section */}
                        <div>
                            <h4>Tags</h4>
                            {formData.tags.map((tag, index) => (
                                <div key={index}>
                                    <input
                                        type="text"
                                        value={tag}
                                        onChange={(e) => handleChange(e, "tags", index)}
                                    />
                                    <button type="button" onClick={() => handleRemoveField("tags", index)}>
                                        Remove
                                    </button>
                                </div>
                            ))}
                            <button type="button" onClick={() => handleAddField("tags")}>
                                Add Tag
                            </button>
                        </div>

                        {/* People Section */}
                        <div>
                            <h4>People</h4>
                            {formData.people.map((person, index) => (
                                <div key={index}>
                                    <input
                                        type="text"
                                        value={person}
                                        onChange={(e) => handleChange(e, "people", index)}
                                    />
                                    <button type="button" onClick={() => handleRemoveField("people", index)}>
                                        Remove
                                    </button>
                                </div>
                            ))}
                            <button type="button" onClick={() => handleAddField("people")}>
                                Add Person
                            </button>
                        </div>

                        <button type="submit">Save Changes</button>
                    </form>
                </>
            )}
        </div>
    );
};

export default EditPictureForm;

File: flutter_frontend/lib/assets/management/forms/editPictureForm.dart
import 'dart:convert';
import 'package:flutter/material.dart';
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'package:http/http.dart' as http;
import 'package:flutter_frontend/static/constants.dart';

class EditPictureForm extends StatefulWidget {
  const EditPictureForm({Key? key}) : super(key: key);

  @override
  State<EditPictureForm> createState() => _EditPictureFormState();
}

class _EditPictureFormState extends State<EditPictureForm> {
  final _formKey = GlobalKey<FormState>();
  final storage = const FlutterSecureStorage();

  bool _loading = true;
  String? _error;
  String? _successMessage;

  final TextEditingController _titleController = TextEditingController();
  final TextEditingController _descriptionController = TextEditingController();

  List<TextEditingController> _tagControllers = [];
  List<TextEditingController> _peopleControllers = [];

  @override
  void initState() {
    super.initState();
    _fetchPicture();
  }

  @override
  void dispose() {
    _titleController.dispose();
    _descriptionController.dispose();
    for (final c in _tagControllers) {
      c.dispose();
    }
    for (final c in _peopleControllers) {
      c.dispose();
    }
    super.dispose();
  }

  Future<void> _fetchPicture() async {
    try {
      final token = await storage.read(key: 'token');
      final serial = await storage.read(key: 'editPictureSerial');

      if (token == null || serial == null) {
        throw Exception('Missing token or picture serial');
      }

      final response = await http.get(
        Uri.parse('$server/get/picture_record/$serial'),
        headers: {'Authorization': token, 'Content-Type': 'application/json'},
      );

      if (response.statusCode != 200) {
        throw Exception('Failed to fetch picture data');
      }

      final data = jsonDecode(response.body)['data'];

      setState(() {
        _titleController.text = data['picture_title'] ?? '';
        _descriptionController.text = data['picture_description'] ?? '';

        _tagControllers = (data['picture_tags'] ?? '')
            .toString()
            .split(',')
            .where((e) => e.isNotEmpty)
            .map((e) => TextEditingController(text: e))
            .toList();

        _peopleControllers = (data['picture_people'] ?? '')
            .toString()
            .split(',')
            .where((e) => e.isNotEmpty)
            .map((e) => TextEditingController(text: e))
            .toList();

        _loading = false;
      });
    } catch (e) {
      setState(() {
        _error = e.toString();
        _loading = false;
      });
    }
  }

  Future<void> _submitForm() async {
    if (!_formKey.currentState!.validate()) return;

    setState(() {
      _error = null;
      _successMessage = null;
    });

    try {
      final token = await storage.read(key: 'token');
      final serial = await storage.read(key: 'editPictureSerial');

      if (token == null || serial == null) {
        throw Exception('Missing token or picture serial');
      }

      final response = await http.patch(
        Uri.parse('$server/update/picture_record/$serial/'),
        headers: {'Authorization': token, 'Content-Type': 'application/json'},
        body: jsonEncode({
          'picture_title': _titleController.text,
          'picture_description': _descriptionController.text,
          'tags': _tagControllers.map((e) => e.text).join(','),
          'people': _peopleControllers.map((e) => e.text).join(','),
        }),
      );

      if (response.statusCode != 200) {
        throw Exception('Failed to update picture');
      }

      setState(() {
        _successMessage = 'Picture details updated successfully!';
      });
    } catch (e) {
      setState(() {
        _error = e.toString();
      });
    }
  }

  Widget _buildDynamicList(
    String title,
    List<TextEditingController> controllers,
    VoidCallback onAdd,
    void Function(int) onRemove,
  ) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Text(
          title,
          style: const TextStyle(fontSize: 16, fontWeight: FontWeight.bold),
        ),
        const SizedBox(height: 8),
        ...controllers.asMap().entries.map((entry) {
          final index = entry.key;
          final controller = entry.value;
          return Row(
            children: [
              Expanded(
                child: TextFormField(
                  controller: controller,
                  validator: (value) {
                    return null;
                  },
                ),
              ),
              IconButton(
                icon: const Icon(Icons.remove_circle),
                onPressed: () => setState(() => onRemove(index)),
              ),
            ],
          );
        }),
        TextButton.icon(
          onPressed: () => setState(onAdd),
          icon: const Icon(Icons.add),
          label: const Text('Add'),
        ),
      ],
    );
  }

  @override
  Widget build(BuildContext context) {
    if (_loading) {
      return const Center(child: CircularProgressIndicator());
    }

    return Padding(
      padding: const EdgeInsets.all(16),
      child: Form(
        key: _formKey,
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            if (_error != null)
              Text(_error!, style: const TextStyle(color: Colors.red)),

            if (_successMessage != null)
              Text(
                _successMessage!,
                style: const TextStyle(color: Colors.green),
              ),

            TextFormField(
              controller: _titleController,
              decoration: const InputDecoration(labelText: 'Title'),
              validator: (value) {
                if (value == null || value.trim().isEmpty) {
                  return 'Please enter a title';
                }
                return null;
              },
            ),

            TextFormField(
              controller: _descriptionController,
              decoration: const InputDecoration(labelText: 'Description'),
              validator: (value) {
                return null;
              },
            ),

            const SizedBox(height: 20),

            _buildDynamicList(
              'Tags',
              _tagControllers,
              () => _tagControllers.add(TextEditingController()),
              (i) => _tagControllers.removeAt(i),
            ),

            const SizedBox(height: 20),

            _buildDynamicList(
              'People',
              _peopleControllers,
              () => _peopleControllers.add(TextEditingController()),
              (i) => _peopleControllers.removeAt(i),
            ),

            const SizedBox(height: 30),

            ElevatedButton(
              onPressed: _submitForm,
              child: const Text('Save Changes'),
            ),
          ],
        ),
      ),
    );
  }
}
